spec/app/ics-029-fee-payment/README.md

```diff
@@ -0,0 +1,433 @@
+  portIdentifier: Identifier,
+  channelIdentifier: Identifier,
+  counterpartyPortIdentifier: Identifier,
+  counterpartyChannelIdentifier: Identifier,
+  version: string) {
+    // remove the prefix and pass the app-specific version to app callback.
+    // otherwise, pass version directly to app callback.
+    feeVersion, appVersion = splitFeeVersion(version)
+    // check that feeVersion is supported
+    if !isSupported(feeVersion) {
+        return error
+    }
+    app.OnChanOpenInit(
+        order,
+        connectionHops,
+        portIdentifier,
+        channelIdentifier,
+        counterpartyPortIdentifier,
+        counterpartyChannelIdentifier,
+        appVersion,
+    )
+}
+
+function OnChanOpenTry(
+  order: ChannelOrder,
+  connectionHops: [Identifier],
+  portIdentifier: Identifier,
+  channelIdentifier: Identifier,
+  counterpartyPortIdentifier: Identifier,
+  counterpartyChannelIdentifier: Identifier,
+  version: string,
+  counterpartyVersion: string) {
+      // select mutually compatible fee version
+      cpFeeVersion, cpAppVersion = splitFeeVersion(counterpartyVersion)
+      feeVersion, appVersion = splitFeeVersion(version)
+      if !isCompatible(cpFeeVersion, feeVersion) {
+          return error
+      }
+      selectFeeVersion(cpFeeVersion, feeVersion)
+
+      // call the underlying applications OnChanOpenTry callback
+      app.OnChanOpenTry(
+          order,
+          connectionHops,
+          portIdentifier,
+          channelIdentifier,
+          counterpartyPortIdentifier,
+          counterpartyChannelIdentifier,
+          cpAppVersion,
+          appVersion,
+      )
+}
+
+function onChanOpenAck(
+  portIdentifier: Identifier,
+  channelIdentifier: Identifier,
+  version: string) {
+      feeVersion, appVersion = splitFeeVersion(version)
+      if !isSupported(feeVersion) {
+          return error
+      }
+
+      // call the underlying applications OnChanOpenAck callback
+      app.OnChanOpenAck(portIdentifier, channelIdentifier, appVersion)
+}
+
+function onChanOpenConfirm(
+  portIdentifier: Identifier,
+  channelIdentifier: Identifier) {
+      // fee middleware performs no-op on ChanOpenConfirm,
+      // just call underlying callback
+      app.onChanOpenConfirm(portIdentifier, channelIdentifier)
+  }
+
+function splitFeeVersion(version: string): []string {
+    if hasPrefix(version, "fee") {
+        splitVersions = split(version,  ":")
+        feeVersion = version[0]
+        appVersion = join(version[1:], ":")
+        // if version has fee prefix
+        // return first split as fee version and the rest of the string as app version
+        return []string{feeVersion, appVersion}
+    }
+    // otherwise return an empty fee version and full version as app version
+    return []string{"", version}
+}
+```
+
+#### Packet Callbacks
+
+```typescript
+function onRecvPacket(packet: Packet, relayer: string): bytes {
+    app_acknowledgement = app.onRecvPacket(packet, relayer)
```

Comment:
I assume `relayer` here is source address? Why we are passing it to the application?

Reply:
The interface is standardized. The Fee middleware is an ICS-26 module, and so is the underlying application. We extended the `OnRecvPacket` to take a relayer to enable applications to implement custom fee logic and to enable fee middleware that implements ICS26